server_test.go

```diff
@@ -556,3 +595,346 @@ func TestRequestSenderPeerIDContext(t *testing.T) {
+	ctxs[0] = context.Background()
+	ctxs[1] = context.Background()
+
+	dests := make([]peer.ID, 2)
+	dests[0] = h1.ID()
+	dests[1] = h2.ID()
+
+	numbers := make(chan Args, 10)
+	quotientss := make([]interface{}, 2)
+	quotients1 := make(chan Quotient, 10)
+	quotients2 := make(chan Quotient, 10)
+	quotientss[0] = quotients1
+	quotientss[1] = quotients2
+
+	numbers <- Args{10, 3}
+	numbers <- Args{10, 3}
+	close(numbers)
+
+	errs := c.MultiStream(ctxs, dests, "Arith", "DivideMyNumbers", numbers, quotientss)
+	for _, err := range errs {
+		if err != nil {
+			t.Fatal(err)
+		}
+	}
+
+	if len(quotients1) != 2 || len(quotients2) != 2 {
+		t.Error("not all channels got responses")
+	}
+
+	for r := range quotients1 {
+		if r.Quo != 3 || r.Rem != 1 {
+			t.Error("wrong result")
+		}
+	}
+	for r := range quotients2 {
+		if r.Quo != 3 || r.Rem != 1 {
+			t.Error("wrong result")
+		}
+	}
+
+}
+
+func TestMultiStreamErrors(t *testing.T) {
+	h1, h2 := makeRandomNodes()
+	defer h1.Close()
+	defer h2.Close()
+
+	s := NewServer(h1, "rpc")
+	s2 := NewServer(h2, "rpc")
+	c := NewClientWithServer(h1, "rpc", s)
+	var arith Arith
+	s.Register(&arith)
+	s2.Register(&arith)
+
+	ctxs := make([]context.Context, 2)
+	ctxs[0] = context.Background()
+	ctxs[1] = context.Background()
+
+	dests := make([]peer.ID, 2)
+	dests[0] = h1.ID()
+	dests[1] = h2.ID()
+
+	numbers := make(chan Args, 10)
+	quotientss := make([]interface{}, 2)
+	quotients1 := make(chan Quotient, 10)
+	quotients2 := make(chan Quotient, 10)
+	quotientss[0] = quotients1
+	quotientss[1] = quotients2
+
+	numbers <- Args{10, 0}
+	numbers <- Args{10, 0}
+	close(numbers)
+
+	errs := c.MultiStream(ctxs, dests, "Arith", "DivideMyNumbers", numbers, quotientss)
+	for _, err := range errs {
+		if err == nil {
+			t.Fatal("expected errors")
+		}
+	}
+}
+
+func TestMultiStreamCancel(t *testing.T) {
+	h1, h2 := makeRandomNodes()
+	defer h1.Close()
+	defer h2.Close()
+
+	s := NewServer(h1, "rpc")
+	s2 := NewServer(h2, "rpc")
+	c := NewClientWithServer(h1, "rpc", s)
+	var arith Arith
+	s.Register(&arith)
+	s2.Register(&arith)
+
+	ctxs := make([]context.Context, 2)
+	var c1, c2 context.CancelFunc
+	ctxs[0], c1 = context.WithTimeout(context.Background(), 500*time.Millisecond)
+	ctxs[1], c2 = context.WithTimeout(context.Background(), 500*time.Millisecond)
+	defer c1()
+	defer c2()
+
+	dests := make([]peer.ID, 2)
+	dests[0] = h1.ID()
+	dests[1] = h2.ID()
+
+	numbers := make(chan Args, 10)
+	quotientss := make([]interface{}, 2)
+	quotients1 := make(chan Quotient, 10)
+	quotients2 := make(chan Quotient, 10)
+	quotientss[0] = quotients1
+	quotientss[1] = quotients2
+
+	numbers <- Args{10, 2}
+	numbers <- Args{1234, 3}
+	close(numbers)
+
+	errs := c.MultiStream(ctxs, dests, "Arith", "DivideMyNumbers", numbers, quotientss)
+	for i, err := range errs {
+		if err == nil {
+			t.Fatal("expected errors")
+			continue
+		}
+		if err.Error() != ctxs[i].Err().Error() {
+			t.Error("expected context error")
+		}
+	}
+}
```

Comment:
Additional test for misbehaving clients/methods:

* A misbehaving client cancels the request but does not close the channel. The test should finish.
* A misbehaving method errors but does not close the reply channel. The test should finish.